File: features/transaction/checkout/src/main/java/com/tokopedia/checkout/view/viewholder/ShipmentCartItemViewHolder.java
package com.tokopedia.checkout.view.viewholder;

import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.text.TextUtils;
import android.view.View;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.flexbox.FlexboxLayout;
import com.tokopedia.abstraction.common.utils.image.ImageHandler;
import com.tokopedia.checkout.R;
import com.tokopedia.checkout.utils.WeightFormatterUtil;
import com.tokopedia.design.utils.CurrencyFormatUtil;
import com.tokopedia.logisticcart.shipping.model.CartItemModel;
import com.tokopedia.purchase_platform.common.utils.Utils;
import com.tokopedia.unifycomponents.ticker.Ticker;
import com.tokopedia.unifyprinciples.Typography;

import java.util.List;

/**
 * @author Aghny A. Putra on 02/03/18
 */

public class ShipmentCartItemViewHolder extends RecyclerView.ViewHolder {

    private static final int IMAGE_ALPHA_DISABLED = 128;
    private static final int IMAGE_ALPHA_ENABLED = 255;

    private ShipmentItemListener shipmentItemListener;

    private ImageView mIvProductImage;
    private Typography mTvProductName;
    private Typography mTvProductPrice;
    private Typography mTvProductOriginalPrice;
    private Typography mTvProductCountAndWeight;
    private TextView mTvOptionalNoteToSeller;
    private RelativeLayout mRlPurchaseProtection;
    private TextView mTvPPPLinkText;
    private TextView mTvPPPPrice;
    private TextView mTvPPPMore;
    private CheckBox mCbPPP;
    private CheckBox mCbPPPDisabled;
    private LinearLayout mLlShippingWarningContainer;
    private View mSeparatorMultipleProductSameStore;
    private TextView tvErrorShipmentItemTitle;
    private TextView tvErrorShipmentItemDescription;
    private Ticker productTicker;
    private Typography mTextVariant;
    private FlexboxLayout mLayoutProductInfo;

    public ShipmentCartItemViewHolder(View itemView) {
        super(itemView);

        mIvProductImage = itemView.findViewById(R.id.iv_product_image);
        mTvProductName = itemView.findViewById(R.id.tv_product_name);
        mTvProductPrice = itemView.findViewById(R.id.tv_product_price);
        mTvProductOriginalPrice = itemView.findViewById(R.id.tv_product_original_price);
        mTvProductCountAndWeight = itemView.findViewById(R.id.tv_item_count_and_weight);
        mTvOptionalNoteToSeller = itemView.findViewById(R.id.tv_optional_note_to_seller);
        mRlPurchaseProtection = itemView.findViewById(R.id.rlayout_purchase_protection);
        mTvPPPLinkText = itemView.findViewById(R.id.text_link_text);
        mTvPPPPrice = itemView.findViewById(R.id.text_price_per_product);
        mTvPPPMore = itemView.findViewById(R.id.text_ppp_more);
        mCbPPP = itemView.findViewById(R.id.checkbox_ppp);
        mCbPPPDisabled = itemView.findViewById(R.id.checkbox_ppp_disabled);
        mLlShippingWarningContainer = itemView.findViewById(R.id.ll_shipping_warning_container);
        mSeparatorMultipleProductSameStore = itemView.findViewById(R.id.v_separator_multiple_product_same_store);
        tvErrorShipmentItemTitle = itemView.findViewById(R.id.tv_error_shipment_item_title);
        tvErrorShipmentItemDescription = itemView.findViewById(R.id.tv_error_shipment_item_description);
        productTicker = itemView.findViewById(R.id.product_ticker);
        mTextVariant = itemView.findViewById(R.id.text_variant);
        mLayoutProductInfo = itemView.findViewById(R.id.layout_product_info);
    }

    public void bindViewHolder(CartItemModel cartItem, ShipmentItemListener listener) {
        shipmentItemListener = listener;
        if (cartItem.isError()) {
            showShipmentWarning(cartItem);
        } else {
            hideShipmentWarning();
        }
        mSeparatorMultipleProductSameStore.setVisibility(View.VISIBLE);
        ImageHandler.LoadImage(mIvProductImage, cartItem.getImageUrl());
        mTvProductName.setText(cartItem.getName());
        mTvProductCountAndWeight.setText(String.format(mTvProductCountAndWeight.getContext()
                        .getString(R.string.iotem_count_and_weight_format),
                String.valueOf(cartItem.getQuantity()),
                WeightFormatterUtil.getFormattedWeight(cartItem.getWeight(), cartItem.getQuantity())));
        if (!TextUtils.isEmpty(cartItem.getVariant())) {
            mTextVariant.setText(cartItem.getVariant());
            mTextVariant.setVisibility(View.VISIBLE);
        } else {
            mTextVariant.setVisibility(View.GONE);
        }
        renderProductPrice(cartItem);
        renderNotesToSeller(cartItem);
        renderPurchaseProtection(cartItem);
        renderProductTicker(cartItem);
        renderProductProperties(cartItem);
    }

    private void renderProductProperties(CartItemModel cartItemModel) {
        List<String> productInformationList = cartItemModel.getProductInformation();
        if (productInformationList != null && !productInformationList.isEmpty()) {
            for (int i = 0; i < productInformationList.size(); i++) {
                Typography productInfo = new Typography(itemView.getContext());
                productInfo.setTextColor(ContextCompat.getColor(itemView.getContext(), com.tokopedia.unifyprinciples.R.color.Neutral_N700_68));
                productInfo.setType(Typography.SMALL);
                if (mLayoutProductInfo.getChildCount() > 0) {
                    productInfo.setText(", " + productInformationList.get(i));
                } else {
                    productInfo.setText(productInformationList.get(i));
                }
                mLayoutProductInfo.addView(productInfo);
            }
            mLayoutProductInfo.setVisibility(View.VISIBLE);
        } else {
            mLayoutProductInfo.setVisibility(View.GONE);
        }
    }

    private void renderProductPrice(CartItemModel cartItem) {
        mTvProductPrice.setText(Utils.removeDecimalSuffix(CurrencyFormatUtil.convertPriceValueToIdrFormat(
                (long) cartItem.getPrice(), false)));
        int dp4 = mTvProductPrice.getResources().getDimensionPixelOffset(R.dimen.dp_4);
        int dp10 = mTvProductPrice.getResources().getDimensionPixelOffset(R.dimen.dp_10);
        if (cartItem.getOriginalPrice() > 0) {
            mTvProductPrice.setPadding(dp4, dp4, 0, 0);
            mTvProductOriginalPrice.setText(Utils.removeDecimalSuffix(CurrencyFormatUtil.convertPriceValueToIdrFormat((long) cartItem.getOriginalPrice(), false)));
            mTvProductOriginalPrice.setPaintFlags(mTvProductOriginalPrice.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
            mTvProductOriginalPrice.setVisibility(View.VISIBLE);
        } else {
            mTvProductPrice.setPadding(dp10, dp4, 0, 0);
            mTvProductOriginalPrice.setVisibility(View.GONE);
        }
    }

    private void renderNotesToSeller(CartItemModel cartItem) {
        if (!TextUtils.isEmpty(cartItem.getNoteToSeller())) {
            mTvOptionalNoteToSeller.setText(cartItem.getNoteToSeller());
            mTvOptionalNoteToSeller.setVisibility(View.VISIBLE);
        } else {
            mTvOptionalNoteToSeller.setVisibility(View.GONE);
        }
    }

    private void renderPurchaseProtection(CartItemModel cartItem) {
        mRlPurchaseProtection.setVisibility(cartItem.isProtectionAvailable() ? View.VISIBLE : View.GONE);
        if (cartItem.isProtectionAvailable()) {
            mTvPPPMore.setText(cartItem.getProtectionLinkText());
            mTvPPPMore.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    shipmentItemListener.navigateToWebView(cartItem.getProtectionLinkUrl());
                }
            });
            mTvPPPLinkText.setText(cartItem.getProtectionTitle());
            mTvPPPPrice.setText(cartItem.getProtectionSubTitle());


            if (cartItem.isProtectionCheckboxDisabled()) {
                mCbPPP.setVisibility(View.GONE);
                mCbPPPDisabled.setVisibility(View.VISIBLE);
                mCbPPPDisabled.setChecked(true);
                mCbPPPDisabled.setClickable(false);
            } else {
                mCbPPPDisabled.setVisibility(View.GONE);
                mCbPPP.setVisibility(View.VISIBLE);
                mCbPPP.setChecked(cartItem.isProtectionOptIn());
                mCbPPP.setClickable(true);
                mCbPPP.setOnCheckedChangeListener((compoundButton, checked) -> shipmentItemListener.notifyOnPurchaseProtectionChecked(checked, getAdapterPosition() + 1));
            }
        }
    }

    private void renderProductTicker(CartItemModel cartItemModel) {
        if (cartItemModel.isShowTicker() && !TextUtils.isEmpty(cartItemModel.getTickerMessage())) {
            productTicker.setVisibility(View.VISIBLE);
            productTicker.setTextDescription(cartItemModel.getTickerMessage());
        } else productTicker.setVisibility(View.GONE);
    }

    private void showShipmentWarning(CartItemModel cartItemModel) {
        if (!TextUtils.isEmpty(cartItemModel.getErrorMessage())) {
            tvErrorShipmentItemTitle.setText(cartItemModel.getErrorMessage());
            tvErrorShipmentItemTitle.setVisibility(View.VISIBLE);
            mLlShippingWarningContainer.setVisibility(View.VISIBLE);
            if (!TextUtils.isEmpty(cartItemModel.getErrorMessageDescription())) {
                tvErrorShipmentItemDescription.setText(cartItemModel.getErrorMessageDescription());
                tvErrorShipmentItemDescription.setVisibility(View.VISIBLE);
            } else {
                tvErrorShipmentItemDescription.setVisibility(View.GONE);
            }
        } else {
            mLlShippingWarningContainer.setVisibility(View.GONE);
        }
        disableItemView();
    }

    private void hideShipmentWarning() {
        mLlShippingWarningContainer.setVisibility(View.GONE);
        enableItemView();
    }

    private void disableItemView() {
        int colorGreyNonActiveText = ContextCompat.getColor(mTvProductName.getContext(), R.color.grey_nonactive_text);
        mTvProductName.setTextColor(colorGreyNonActiveText);
        mTvProductPrice.setTextColor(colorGreyNonActiveText);
        mTvProductOriginalPrice.setTextColor(colorGreyNonActiveText);
        mTvOptionalNoteToSeller.setTextColor(colorGreyNonActiveText);
        mTvProductCountAndWeight.setTextColor(colorGreyNonActiveText);
        mTextVariant.setTextColor(colorGreyNonActiveText);
        setImageFilterGrayScale();
    }

    private void setImageFilterGrayScale() {
        ColorMatrix matrix = new ColorMatrix();
        matrix.setSaturation(0);
        ColorMatrixColorFilter disabledColorFilter = new ColorMatrixColorFilter(matrix);
        mIvProductImage.setColorFilter(disabledColorFilter);
        mIvProductImage.setImageAlpha(IMAGE_ALPHA_DISABLED);
    }

    private void enableItemView() {
        mTvProductName.setTextColor(ContextCompat.getColor(mTvProductName.getContext(), com.tokopedia.unifyprinciples.R.color.Neutral_N700_96));
        mTextVariant.setTextColor(ContextCompat.getColor(mTextVariant.getContext(), com.tokopedia.unifyprinciples.R.color.Neutral_N700_68));
        mTvProductPrice.setTextColor(ContextCompat.getColor(mTvProductPrice.getContext(), com.tokopedia.unifyprinciples.R.color.Neutral_N700_96));
        mTvProductOriginalPrice.setTextColor(ContextCompat.getColor(mTvProductOriginalPrice.getContext(), com.tokopedia.unifyprinciples.R.color.Neutral_N700_68));
        mTvProductCountAndWeight.setTextColor(ContextCompat.getColor(mTvProductCountAndWeight.getContext(), com.tokopedia.unifyprinciples.R.color.Neutral_N700_68));
        mTvOptionalNoteToSeller.setTextColor(ContextCompat.getColor(mTvOptionalNoteToSeller.getContext(), com.tokopedia.unifyprinciples.R.color.Neutral_N700_96));
        setImageFilterNormal();
    }

    private void setImageFilterNormal() {
        mIvProductImage.setColorFilter(null);
        mIvProductImage.setImageAlpha(IMAGE_ALPHA_ENABLED);
    }

    public interface ShipmentItemListener {
        void notifyOnPurchaseProtectionChecked(boolean checked, int position);

        void navigateToWebView(String protectionLinkUrl);
    }

}

File: features/transaction/checkout/src/main/java/com/tokopedia/checkout/view/viewholder/ShipmentCartItemViewHolder.kt
package com.tokopedia.checkout.view.viewholder

import android.graphics.Paint
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup.MarginLayoutParams
import android.widget.CompoundButton
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.tokopedia.abstraction.common.utils.image.ImageHandler
import com.tokopedia.checkout.R
import com.tokopedia.checkout.databinding.ItemShipmentProductBinding
import com.tokopedia.checkout.domain.mapper.ShipmentMapper
import com.tokopedia.checkout.utils.WeightFormatterUtil
import com.tokopedia.kotlin.extensions.view.dpToPx
import com.tokopedia.kotlin.extensions.view.gone
import com.tokopedia.kotlin.extensions.view.invisible
import com.tokopedia.kotlin.extensions.view.show
import com.tokopedia.kotlin.extensions.view.visible
import com.tokopedia.logisticcart.shipping.model.CartItemModel
import com.tokopedia.purchase_platform.common.databinding.ItemProductInfoAddOnBinding
import com.tokopedia.purchase_platform.common.feature.gifting.data.model.AddOnWordingModel
import com.tokopedia.purchase_platform.common.feature.gifting.view.ButtonGiftingAddOnView
import com.tokopedia.purchase_platform.common.utils.removeDecimalSuffix
import com.tokopedia.unifyprinciples.Typography
import com.tokopedia.unifyprinciples.Typography.Companion.SMALL
import com.tokopedia.utils.currency.CurrencyFormatUtil
import java.util.*

class ShipmentCartItemViewHolder(
    itemView: View,
    private val listener: Listener? = null
) : RecyclerView.ViewHolder(itemView) {

    companion object {

        @JvmField
        val LAYOUT = R.layout.item_shipment_product

        private const val VIEW_ALPHA_ENABLED = 1.0f
        private const val VIEW_ALPHA_DISABLED = 0.5f
    }

    private val binding: ItemShipmentProductBinding = ItemShipmentProductBinding.bind(itemView)

    fun bind(
        cartItem: CartItemModel
    ) {
        renderShopInfo(cartItem)
        renderError(cartItem)
        renderItem(cartItem)
        renderProductPrice(cartItem)
        renderNotesToSeller(cartItem)
        renderPurchaseProtection(cartItem)
        renderProductTicker(cartItem)
        renderProductProperties(cartItem)
        val isFirstItem = cartItem.cartItemPosition == 0
        renderBundlingInfo(cartItem, isFirstItem)
        renderAddOnProductLevel(cartItem, cartItem.addOnOrderLevelModel)
    }

    private fun renderShopInfo(cartItem: CartItemModel) {
        if (cartItem.shouldShowShopInfo) {
            binding.tvShopName.text = cartItem.shopName
            if (cartItem.shopTypeInfoData.shopBadge.isNotEmpty()) {
                binding.imgShopBadge.setImageUrl(cartItem.shopTypeInfoData.shopBadge)
                binding.imgShopBadge.visible()
                binding.imgShopBadge.contentDescription = itemView.context.getString(
                    com.tokopedia.purchase_platform.common.R.string.pp_cd_image_shop_badge_with_shop_type,
                    cartItem.shopTypeInfoData.title.lowercase(
                        Locale.getDefault()
                    )
                )
            } else {
                binding.imgShopBadge.gone()
            }
            binding.productShopInfo.visible()
        } else {
            binding.productShopInfo.gone()
        }
    }

    private fun renderError(cartItem: CartItemModel) {
        if (cartItem.isError) {
            binding.productBundlingInfo.alpha = VIEW_ALPHA_DISABLED
            binding.llFrameItemProductContainer.alpha = VIEW_ALPHA_DISABLED
            binding.checkboxPpp.isEnabled = false
            binding.iconTooltip.isClickable = false
        } else {
            binding.productBundlingInfo.alpha = VIEW_ALPHA_ENABLED
            binding.llFrameItemProductContainer.alpha = VIEW_ALPHA_ENABLED
            binding.checkboxPpp.isEnabled = true
            binding.iconTooltip.isClickable = true
        }
    }

    private fun renderItem(cartItem: CartItemModel) {
        if (cartItem.isError) {
            showShipmentWarning(cartItem)
        } else {
            hideShipmentWarning()
        }
        binding.ivProductImage.setImageUrl(cartItem.imageUrl)
        binding.tvProductName.text = cartItem.name
        binding.tvItemCountAndWeight.text = String.format(
            binding.tvItemCountAndWeight.context.getString(R.string.iotem_count_and_weight_format),
            cartItem.quantity.toString(),
            WeightFormatterUtil.getFormattedWeight(cartItem.weight, cartItem.quantity)
        )
        if (!TextUtils.isEmpty(cartItem.variant)) {
            binding.textVariant.text = cartItem.variant
            binding.textVariant.visible()
        } else {
            binding.textVariant.gone()
        }
    }

    private fun renderProductProperties(cartItemModel: CartItemModel) {
        binding.layoutProductInfo.removeAllViews()
        if (cartItemModel.productInformation.isNotEmpty()) {
            cartItemModel.productInformation.forEach { productInformation ->
                val productInfo = Typography(itemView.context)
                productInfo.setTextColor(
                    ContextCompat.getColor(
                        itemView.context,
                        com.tokopedia.unifyprinciples.R.color.Unify_NN950_68
                    )
                )
                productInfo.setType(SMALL)
                if (binding.layoutProductInfo.childCount > 0) {
                    productInfo.text = ", $productInformation"
                } else {
                    productInfo.text = productInformation
                }
                binding.layoutProductInfo.addView(productInfo)
            }
            binding.layoutProductInfo.visibility = View.VISIBLE
        } else {
            binding.layoutProductInfo.visibility = View.GONE
        }
        renderEthicalDrugsProperty(cartItemModel)
    }

    private fun renderEthicalDrugsProperty(cartItemModel: CartItemModel) {
        if (cartItemModel.ethicalDrugDataModel.needPrescription) {
            val ethicalDrugView: View = createProductInfoTextWithIcon(cartItemModel)
            if (binding.layoutProductInfo.childCount > 0) {
                ethicalDrugView.setPadding(
                    4.dpToPx(ethicalDrugView.resources.displayMetrics),
                    0,
                    0,
                    0
                )
            }
            binding.layoutProductInfo.addView(ethicalDrugView)
            binding.layoutProductInfo.visibility = View.VISIBLE
        }
    }

    private fun createProductInfoTextWithIcon(cartItemModel: CartItemModel): LinearLayout {
        val propertyLayoutWithIcon = LinearLayout(itemView.context)
        propertyLayoutWithIcon.orientation = LinearLayout.HORIZONTAL
        ItemProductInfoAddOnBinding.inflate(LayoutInflater.from(itemView.context), null, false)
        val propertiesBinding = ItemProductInfoAddOnBinding.inflate(LayoutInflater.from(itemView.context), null, false)
        if (!TextUtils.isEmpty(cartItemModel.ethicalDrugDataModel.iconUrl)) {
            ImageHandler.loadImageWithoutPlaceholderAndError(
                propertiesBinding.ppIvProductInfoAddOn,
                cartItemModel.ethicalDrugDataModel.iconUrl
            )
        }
        if (!TextUtils.isEmpty(cartItemModel.ethicalDrugDataModel.text)) {
            propertiesBinding.ppLabelProductInfoAddOn.text = cartItemModel.ethicalDrugDataModel.text
        }
        propertyLayoutWithIcon.addView(propertiesBinding.root)
        return propertyLayoutWithIcon
    }

    private fun renderProductPrice(cartItem: CartItemModel) {
        binding.tvProductPrice.text = CurrencyFormatUtil
            .convertPriceValueToIdrFormat(cartItem.price.toLong(), false)
            .removeDecimalSuffix()
        val dp4 =
            binding.tvProductPrice.resources.getDimensionPixelOffset(com.tokopedia.abstraction.R.dimen.dp_4)
        if (cartItem.originalPrice > 0) {
            binding.tvProductPrice.setPadding(0, dp4, 0, 0)
            binding.tvProductOriginalPrice.setPadding(0, dp4, 0, 0)
            binding.tvProductOriginalPrice.text = CurrencyFormatUtil
                .convertPriceValueToIdrFormat(cartItem.originalPrice.toLong(), false)
                .removeDecimalSuffix()
            binding.tvProductOriginalPrice.paintFlags =
                binding.tvProductOriginalPrice.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
            binding.tvProductOriginalPrice.visibility = View.VISIBLE
        } else {
            binding.tvProductPrice.setPadding(0, dp4, 0, 0)
            binding.tvProductOriginalPrice.visibility = View.GONE
        }
    }

    private fun renderNotesToSeller(cartItem: CartItemModel) {
        if (!TextUtils.isEmpty(cartItem.noteToSeller)) {
            binding.tvOptionalNoteToSeller.text = cartItem.noteToSeller
            binding.tvOptionalNoteToSeller.visibility = View.VISIBLE
        } else {
            binding.tvOptionalNoteToSeller.visibility = View.GONE
        }
    }

    private fun renderPurchaseProtection(cartItem: CartItemModel) {
        binding.rlayoutPurchaseProtection.visibility =
            if (cartItem.isProtectionAvailable && !cartItem.isError) {
                View.VISIBLE
            } else {
                View.GONE
            }
        if (cartItem.isProtectionAvailable && !cartItem.isError) {
            binding.iconTooltip.setOnClickListener {
                listener?.onClickPurchaseProtectionTooltip(
                    cartItem
                )
            }
            binding.textLinkText.text = cartItem.protectionTitle
            binding.textProtectionDesc.text = cartItem.protectionSubTitle
            binding.textItemPerProduct.text = CurrencyFormatUtil
                .convertPriceValueToIdrFormat(cartItem.protectionPricePerProduct.toLong(), false)
                .removeDecimalSuffix()
            binding.checkboxPpp.setOnCheckedChangeListener { _, _ -> }
            if (cartItem.isProtectionCheckboxDisabled) {
                binding.checkboxPpp.isEnabled = false
                binding.checkboxPpp.isChecked = true
                binding.checkboxPpp.skipAnimation()
            } else {
                binding.checkboxPpp.isEnabled = true
                binding.checkboxPpp.isChecked = cartItem.isProtectionOptIn
                binding.checkboxPpp.skipAnimation()
                binding.checkboxPpp.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                    if (bindingAdapterPosition != RecyclerView.NO_POSITION) {
                        val updatedCartItemModel = cartItem.copy(
                            isProtectionOptIn = isChecked
                        )
                        listener?.onCheckPurchaseProtection(bindingAdapterPosition, updatedCartItemModel)
                    }
                }
            }
        }
    }

    private fun renderProductTicker(cartItemModel: CartItemModel) {
        if (cartItemModel.isShowTicker && !TextUtils.isEmpty(cartItemModel.tickerMessage)) {
            binding.productTicker.visibility = View.VISIBLE
            binding.productTicker.setTextDescription(cartItemModel.tickerMessage)
        } else {
            binding.productTicker.visibility = View.GONE
        }
    }

    private fun showShipmentWarning(cartItemModel: CartItemModel) {
        if (!TextUtils.isEmpty(cartItemModel.errorMessage)) {
            if (!TextUtils.isEmpty(cartItemModel.errorMessageDescription)) {
                binding.checkoutTickerProductError.tickerTitle = cartItemModel.errorMessage
                binding.checkoutTickerProductError.setTextDescription(cartItemModel.errorMessageDescription)
            } else {
                binding.checkoutTickerProductError.setTextDescription(cartItemModel.errorMessage)
            }

            if (cartItemModel.isBundlingItem) {
                if (cartItemModel.bundlingItemPosition == ShipmentMapper.BUNDLING_ITEM_HEADER) {
                    binding.checkoutTickerProductError.visible()
                    binding.checkoutTickerProductError.post {
                        binding.checkoutTickerProductError.requestLayout()
                    }
                } else {
                    binding.checkoutTickerProductError.gone()
                }
            } else {
                binding.checkoutTickerProductError.visible()
                binding.checkoutTickerProductError.post {
                    binding.checkoutTickerProductError.requestLayout()
                }
            }
        } else {
            binding.checkoutTickerProductError.gone()
        }

        if (!cartItemModel.isShopError) {
            disableItemView()
        }
    }

    private fun hideShipmentWarning() {
        binding.checkoutTickerProductError.gone()
        enableItemView()
    }

    private fun enableItemView() {
        binding.productBundlingInfo.alpha = VIEW_ALPHA_ENABLED
        binding.llFrameItemProductContainer.alpha = VIEW_ALPHA_ENABLED
    }

    private fun disableItemView() {
        binding.productBundlingInfo.alpha = VIEW_ALPHA_DISABLED
        binding.llFrameItemProductContainer.alpha = VIEW_ALPHA_DISABLED
    }

    private fun renderBundlingInfo(cartItemModel: CartItemModel, isFirstItem: Boolean) {
        val ivProductImageLayoutParams = binding.ivProductImage.layoutParams as MarginLayoutParams
        val tvOptionalNoteToSellerLayoutParams =
            binding.tvOptionalNoteToSeller.layoutParams as MarginLayoutParams
        val productContainerLayoutParams =
            binding.llFrameItemProductContainer.layoutParams as MarginLayoutParams
        val productInfoLayoutParams = binding.rlProductInfo.layoutParams as MarginLayoutParams
        val bottomMargin =
            itemView.resources.getDimensionPixelSize(com.tokopedia.abstraction.R.dimen.dp_8)
        if (cartItemModel.isBundlingItem) {
            if (!TextUtils.isEmpty(cartItemModel.bundleIconUrl)) {
                ImageHandler.loadImage2(
                    binding.imageBundle,
                    cartItemModel.bundleIconUrl,
                    com.tokopedia.utils.R.drawable.ic_loading_placeholder
                )
            }

            ivProductImageLayoutParams.leftMargin =
                itemView.resources.getDimensionPixelSize(com.tokopedia.abstraction.R.dimen.dp_14)
            tvOptionalNoteToSellerLayoutParams.leftMargin =
                itemView.resources.getDimensionPixelSize(com.tokopedia.abstraction.R.dimen.dp_14)
            binding.vBundlingProductSeparator.visibility = View.VISIBLE
            val productImageLayoutParams = binding.ivProductImage.layoutParams as MarginLayoutParams
            val productNameLayoutParams = binding.tvProductName.layoutParams as MarginLayoutParams
            val productMarginTop =
                itemView.resources.getDimensionPixelSize(com.tokopedia.abstraction.R.dimen.dp_12)
            if (cartItemModel.bundlingItemPosition == ShipmentMapper.BUNDLING_ITEM_HEADER) {
                binding.productBundlingInfo.visibility = View.VISIBLE
                productImageLayoutParams.topMargin = 0
                productNameLayoutParams.topMargin = 0
                binding.vSeparatorMultipleProductSameStore.invisible()
            } else {
                binding.productBundlingInfo.visibility = View.GONE
                productImageLayoutParams.topMargin = productMarginTop
                productNameLayoutParams.topMargin = productMarginTop
                binding.vSeparatorMultipleProductSameStore.gone()
            }
            binding.textBundleTitle.text = cartItemModel.bundleTitle
            binding.textBundlePrice.text =
                CurrencyFormatUtil.convertPriceValueToIdrFormat(cartItemModel.bundlePrice, false)
                    .removeDecimalSuffix()
            binding.textBundleSlashPrice.text = CurrencyFormatUtil.convertPriceValueToIdrFormat(
                cartItemModel.bundleOriginalPrice,
                false
            ).removeDecimalSuffix()
            binding.textBundleSlashPrice.paintFlags =
                binding.textBundleSlashPrice.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
            productContainerLayoutParams.bottomMargin =
                if (cartItemModel.bundlingItemPosition == ShipmentMapper.BUNDLING_ITEM_FOOTER) {
                    bottomMargin
                } else {
                    0
                }
            productInfoLayoutParams.bottomMargin =
                if (cartItemModel.bundlingItemPosition == ShipmentMapper.BUNDLING_ITEM_FOOTER) {
                    bottomMargin
                } else {
                    0
                }
        } else {
            ivProductImageLayoutParams.leftMargin = 0
            tvOptionalNoteToSellerLayoutParams.leftMargin = 0
            binding.vBundlingProductSeparator.visibility = View.GONE
            binding.productBundlingInfo.visibility = View.GONE
            if (isFirstItem) {
                binding.vSeparatorMultipleProductSameStore.gone()
            } else {
                binding.vSeparatorMultipleProductSameStore.show()
            }
            productContainerLayoutParams.bottomMargin = bottomMargin
            productInfoLayoutParams.bottomMargin = bottomMargin
        }
        if (cartItemModel.isLastItemInOrder) {
            productContainerLayoutParams.bottomMargin = bottomMargin
            productInfoLayoutParams.bottomMargin = bottomMargin
        }
    }

    private fun renderAddOnProductLevel(
        cartItemModel: CartItemModel,
        addOnWordingModel: AddOnWordingModel
    ) {
        val addOns = cartItemModel.addOnProductLevelModel
        if (addOns.status == 0) {
            binding.llGiftingAddonProductLevel.visibility = View.GONE
        } else {
            with(binding.itemShipmentGiftingAddonProductLevel) {
                if (addOns.status == 1) {
                    buttonGiftingAddonProductLevel.state = ButtonGiftingAddOnView.State.ACTIVE
                } else if (addOns.status == 2) {
                    buttonGiftingAddonProductLevel.state = ButtonGiftingAddOnView.State.INACTIVE
                }
                buttonGiftingAddonProductLevel.title = addOns.addOnsButtonModel.title
                buttonGiftingAddonProductLevel.desc = addOns.addOnsButtonModel.description
                buttonGiftingAddonProductLevel.urlLeftIcon = addOns.addOnsButtonModel.leftIconUrl
                buttonGiftingAddonProductLevel.urlRightIcon = addOns.addOnsButtonModel.rightIconUrl
                buttonGiftingAddonProductLevel.setOnClickListener {
                    listener?.onClickAddOnProductLevel(
                        cartItemModel,
                        addOnWordingModel
                    )
                }
            }
            binding.llGiftingAddonProductLevel.visibility = View.VISIBLE
            listener?.onImpressionAddOnProductLevel(cartItemModel.productId.toString())
        }
    }

    interface Listener {

        fun onCheckPurchaseProtection(position: Int, cartItem: CartItemModel)

        fun onClickPurchaseProtectionTooltip(cartItem: CartItemModel)

        fun onClickAddOnProductLevel(cartItem: CartItemModel, addOnWording: AddOnWordingModel)

        fun onImpressionAddOnProductLevel(productId: String)
    }
}
